### PR TITLE
Add CircuitFunction

### PR DESCRIPTION
**Pull request by arettig**
This adds `CircuitFunction`, which maps from a set of parameters to a
circuit.`CircuitFunction` will be used by the planned `CallFunction`
operation in a future PR. This is an experimental Cirq 2.0 feature.

### CODE REVIEW

File: cirq-core/cirq/circuits/circuit_function.py
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:1]: It looks like this is missing a `_resolve_parameters_`. Is that intentional or being handled some other way? Without that method, I believe passing a `CircuitFunction` to `cirq.resolve_parameters` will be a silent no-op and leave `CircuitFunction` unresolved.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        self._qubits = self._circuit.all_qubits()
+
+        if function_params is None:
+            circuit_symbols = protocols.parameter_symbols(self._circuit)
+            self._function_params = tuple(sorted(circuit_symbols, key=lambda s: s.name))
+        else:
+            self._function_params = tuple(function_params)
+            if not all(isinstance(p, sympy.Symbol) for p in self._function_params):
+                raise TypeError("All parameters must be sympy Symbols.")
+            if len(set(self._function_params)) != len(self._function_params):
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:79]: I believe `cirq.ParamResolver` converts keys based on string names. However, in SymPy, two symbols can have the same name and still be different due to other properties:

```python
s1 = sympy.Symbol('x', positive=True)
s2 = sympy.Symbol('x', positive=False)
s1 == s2  # False
```
The length comparison using the objects themselves will not detect objects that should be considered the same as far as ParamResolver is concerned. The above should probably compare parameter names, like

```python
param_names = [p.name for p in self._function_params]
if len(set(param_names)) != len(param_names): ....
```

File: cirq-core/cirq/circuits/circuit_function.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    def _from_json_dict_(
+        cls,
+        name: str,
+        circuit: cirq.FrozenCircuit,
+        function_params: Sequence[sympy.Symbol],
+        **kwargs,
+    ) -> CircuitFunction:
+        return cls(name=name, circuit=circuit, function_params=function_params)
+
+    def __call__(self, *call_params: cirq.TParamVal) -> cirq.AbstractCircuit:
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:123]: It may be good to also allow keyword arguments `**kwargs` in calls.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        """Call the circuit function with given parameters values."""
+        if len(call_params) != len(self.function_params):
+            raise TypeError(
+                f"CircuitFunction {self.name} called with the wrong number of parameters."
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:127]: Following Python best practices, it's good for exceptions to report both expected and received argument counts.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 2 above)
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:123]: The return type probably should be `FrozenCircuit`.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    @property
+    def circuit(self) -> FrozenCircuit:
+        return self._circuit
+
+    @property
+    def function_params(self) -> tuple[sympy.Symbol, ...]:
+        return self._function_params
+
+    @property
+    def qubits(self) -> frozenset[cirq.Qid]:
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:95]: I'm not sure this is the correct return type – please check.

Also, should this define an `all_qubits(self)` too?

File: cirq-core/cirq/circuits/circuit_function_test.py
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:1]: Some additional tests that may be worth adding:

* `cirq.resolve_parameters` on `CircuitFunction`
* More `__call__` parameter checks, like when the circuit contains non-function parameters, also symbolic expressions or other symbols as values.

**Review by mhucka — CHANGES_REQUESTED** [review]
Thanks for this great work!

Here are some initial review findings.

File: cirq-core/cirq/circuits/circuit_function.py
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py:1]: This was intentional. There will be one `CircuitFunction` instance that is used by multiple `CallFunction` operations and I don't want `resolve_parameters` to change all locations that this function is used throughout the circuit. I will handle parameter resolution within the `CallFunction` operation when that is added.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 1 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Good catch, fixed.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 3 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Good idea, fixed.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 2 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Yep, fixed.

Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Good idea, added kwargs.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 4 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: I got rid of the `qubits` property and changed this to an `all_qubits` method instead to better align with the `Circuit` class rather than the `CircuitOperation` class (since this is not an Operation).

File: cirq-core/cirq/circuits/circuit_function_test.py
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function_test.py:1]: All good ideas thanks. I added tests to make sure `resolve_parameters` does not affect the `CircuitFunction` as well as tests for a variety of different `__call__` use cases.

**Comment by dstrain115** [comment]
LGTM.  Will wait on mhucka's review for approval.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        Args:
+            name: The name of the function.
+            circuit: The circuit body. If mutable, it will be frozen.
+            function_params: Optional ordered sequence of parameters to the
+                `CircuitFunction`. If None, defaults to all symbols in `circuit`.
+
+        Raises:
+            TypeError: If `name` is not a str, `circuit` is not an `AbstractCircuit`,
+                or any item in `function_params` is not a `sympy.Symbol`.
+            ValueError: If `name` is empty.
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:58]: ValueError is also raised when the CircuitFunction has duplicate parameters (line 80).

File: cirq-core/cirq/circuits/circuit_function.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+    @property
+    def circuit(self) -> FrozenCircuit:
+        return self._circuit
+
+    @property
+    def function_params(self) -> tuple[sympy.Symbol, ...]:
+        return self._function_params
+
+    def all_qubits(self) -> frozenset[cirq.Qid]:
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:94]: Should have a docstring.

```suggestion
    def all_qubits(self) -> frozenset[cirq.Qid]:
        """Returns the set of all qubits in the circuit."""
```

File: cirq-core/cirq/circuits/circuit_function.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        cls,
+        name: str,
+        circuit: cirq.FrozenCircuit,
+        function_params: Sequence[sympy.Symbol],
+        **kwargs,
+    ) -> CircuitFunction:
+        return cls(name=name, circuit=circuit, function_params=function_params)
+
+    def __call__(self, *args: cirq.TParamVal, **kwargs: cirq.TParamVal) -> cirq.FrozenCircuit:
+        """Call the circuit function with given parameters values."""
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:123]:
```suggestion
        """Call the circuit function with the given parameter values."""
```

File: cirq-core/cirq/circuits/circuit_function.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    from cirq.circuits.frozen_circuit import FrozenCircuit
+
+
+@value.value_equality
+class CircuitFunction:
+    """A function that maps from a given set of parameters to a circuit.
+
+    Note: This is an experimental class designed as part of a prototype
+       for Cirq 2.0. The interface for this class is subject to change
+       between versions.
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:38]: Hanging indent should have 4 spaces:

```suggestion
        for Cirq 2.0. The interface for this class is subject to change
        between versions.
```

File: cirq-core/cirq/circuits/circuit_function_test.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    cf = cirq.CircuitFunction("test_function", c)
+    assert cf.function_params == (theta, x)
+
+    # parameter not exposed, vqid exposed
+    cf = cirq.CircuitFunction("test_function", c, function_params=[x])
+    assert cf.function_params == (x,)
+    assert cirq.parameter_names(cf) == {theta.name}
+    assert cirq.is_parameterized(cf)
+
+    # parameter not exposed, vqid not exposed
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:107]: Style.

```suggestion
    # Parameter not exposed, vqid not exposed.
```

File: cirq-core/cirq/circuits/circuit_function_test.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+    # parameter exposed, vqid exposed
+    cf = cirq.CircuitFunction("test_function", c, function_params=[theta, x])
+    assert cf.function_params == (theta, x)
+    assert cirq.parameter_names(cf) == set()
+    assert not cirq.is_parameterized(cf)
+
+    cf = cirq.CircuitFunction("test_function", c)
+    assert cf.function_params == (theta, x)
+
+    # parameter not exposed, vqid exposed
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:101]: Style.

```suggestion
    # Parameter not exposed, vqid exposed.
```

File: cirq-core/cirq/circuits/circuit_function_test.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    c = cirq.Circuit(cirq.X(q) ** theta)
+    cf = cirq.CircuitFunction("test_function", c, function_params=[theta])
+    assert cf.name == "test_function"
+    assert cf.circuit == c.freeze()
+    assert cf.function_params == (theta,)
+    assert cirq.parameter_names(cf) == {x.name}
+    assert cirq.is_parameterized(cf)
+    assert cf.all_qubits() == frozenset((q,))
+
+    # parameter exposed, vqid exposed
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:92]:
```suggestion
    # Parameter exposed, vqid exposed.
```

File: cirq-core/cirq/circuits/circuit_function_test.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    assert cf.circuit == c.freeze()
+    assert cf.function_params == (theta,)
+    assert cirq.parameter_names(cf) == set()
+    assert not cirq.is_parameterized(cf)
+    assert cf.all_qubits() == frozenset((q,))
+
+    cf = cirq.CircuitFunction("test_function", c)
+    assert cf.function_params == (theta,)
+
+    # parameter not exposed
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:56]:
```suggestion
    # Parameter not exposed.
```

File: cirq-core/cirq/circuits/circuit_function_test.py [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+    c = cirq.Circuit(cirq.X(q))
+    cf = cirq.CircuitFunction("test_function", c)
+    assert cf.name == "test_function"
+    assert cf.circuit == c.freeze()
+    assert cf.function_params == ()
+    assert cirq.parameter_names(cf) == set()
+    assert not cirq.is_parameterized(cf)
+    assert cf.all_qubits() == frozenset((q,))
+
+    # expose nonfunctional parameter
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:32]: Style.

```suggestion
    # Expose nonfunctional parameter.
```

File: cirq-core/cirq/circuits/circuit_function_test.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    theta = sympy.Symbol('theta')
+    cf = cirq.CircuitFunction("test_function", c, function_params=[theta])
+    assert cf.function_params == (theta,)
+    assert cirq.parameter_names(cf) == set()
+    assert not cirq.is_parameterized(cf)
+
+
+def test_construction_basic() -> None:
+    # parameter exposed
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:41]: Style.

```suggestion
    # Parameter exposed.
```

File: cirq-core/cirq/circuits/circuit_function_test.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+
+    with pytest.raises(TypeError, match='All parameters must be sympy Symbols.'):
+        _ = cirq.CircuitFunction("test_function", c, function_params=[5])
+
+    with pytest.raises(ValueError, match='CircuitFunctions may not have duplicate parameters.'):
+        _ = cirq.CircuitFunction("test_function", c, function_params=[theta, theta])
+
+
+def test_construction_vqid() -> None:
+    # parameter exposed, vqid not exposed
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:79]:
```suggestion
    # Parameter exposed, vqid not exposed.
```

File: cirq-core/cirq/circuits/circuit_function_test.py
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:1]: Some more edge cases to test, maybe:

* Purely keyword call: calling cf(x=2, theta=3) with all args being keyword args.

* Out-of-order purely keyword call: calling cf(theta=3, x=2) with all args being keyword args but out of order.
 
* Zero-parameter call: calling an unparameterized `CircuitFunction: cf()` with no arguments.
 
* Supplying too many positional arguments.

* Having duplicate symbol names but with distinct SymPy properties.

* Checking that argument order matters in equality tests. E.g., adding a group with reversed parameter order (`function_params=[theta, x]` vs `function_params=[x, theta]`) to `test_eq` to verify that argument ordering matters to value equality.

* `cirq.testing.assert_consistent_resolve_parameters(cf)`

File: cirq-core/cirq/circuits/circuit_function.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+        param_dict = dict(zip(self._function_params, args)) | kwargs
+        positional_param_names = {p.name for p in self._function_params[: len(args)]}
+        expected_param_names = {p.name for p in self._function_params}
+
+        if not positional_param_names.isdisjoint(kwargs.keys()):
+            raise TypeError(f"CircuitFunction {self.name} called with duplicate parameters.")
+        if len(args) + len(kwargs) != len(self._function_params):
+            raise TypeError(
+                f"CircuitFunction {self.name} takes {len(self.function_params)}"
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:133]: Shouldn't that be `self._function_params`?
```suggestion
                f"CircuitFunction {self.name} takes {len(self._function_params)}"
```

File: cirq-core/cirq/circuits/circuit_function.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+        else:
+            self._function_params = tuple(function_params)
+            if not all(isinstance(p, sympy.Symbol) for p in self._function_params):
+                raise TypeError("All parameters must be sympy Symbols.")
+            param_names = [p.name for p in self._function_params]
+            if len(set(param_names)) != len(param_names):
+                raise ValueError("CircuitFunctions may not have duplicate parameters.")
+
+    @property
+    def name(self) -> str:
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:83]: Should have a docstring.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+        return self._name
+
+    @property
+    def circuit(self) -> FrozenCircuit:
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:87]: Docstring.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+        return self._circuit
+
+    @property
+    def function_params(self) -> tuple[sympy.Symbol, ...]:
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:91]: Docstring.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 7 above)
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:123]: The docstring lacks sections for `Args:`, `Returns:`, and `Raises:`.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+        circuit: cirq.AbstractCircuit,
+        function_params: Sequence[sympy.Symbol] | None = None,
+    ) -> None:
+        """Initializes a CircuitFunction.
+
+        Args:
+            name: The name of the function.
+            circuit: The circuit body. If mutable, it will be frozen.
+            function_params: Optional ordered sequence of parameters to the
+                `CircuitFunction`. If None, defaults to all symbols in `circuit`.
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:53]: This could use some more detail, in particular that the default is all symbols sorted alphabetically by name.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+
+    Note: This is an experimental class designed as part of a prototype
+       for Cirq 2.0. The interface for this class is subject to change
+       between versions.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        circuit: cirq.AbstractCircuit,
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:44]: To be consistent with the rest of the code:
```suggestion
        circuit: AbstractCircuit,
```

File: cirq-core/cirq/circuits/circuit_function_test.py [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+    )
+
+
+def test_call() -> None:
+    x = sympy.Symbol('x')
+    qx = cirq.VariableLineQid(x)
+    theta = sympy.Symbol('theta')
+    cx = cirq.Circuit(cirq.X(qx) ** theta)
+    cf = cirq.CircuitFunction("test_function", cx, function_params=[x, theta])
+    assert cf(2, 3) == cirq.Circuit(cirq.X(cirq.q(2)) ** 3)
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function_test.py:172]: In addition, we should add checks on the return types. Since `FrozenCircuit == Circuit` in Cirq, the above does not verify the return type. An explicit assert `isinstance(cf(2, 3), cirq.FrozenCircuit)` would be needed.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+        circuit: cirq.FrozenCircuit,
+        function_params: Sequence[sympy.Symbol],
+        **kwargs,
+    ) -> CircuitFunction:
+        return cls(name=name, circuit=circuit, function_params=function_params)
+
+    def __call__(self, *args: cirq.TParamVal, **kwargs: cirq.TParamVal) -> cirq.FrozenCircuit:
+        """Call the circuit function with given parameters values."""
+
+        param_dict = dict(zip(self._function_params, args)) | kwargs
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:125]: This is being constructed before args and kwargs are validated. If the lengths of those are different, `zip` will silently drop the extras.

Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:125]: Also, don't those two dicts have different types? If so, `param_dict` will be heterogeneous. Not necessarily an error, but may lead to unexpected results somewhere.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+        expected_param_names = {p.name for p in self._function_params}
+
+        if not positional_param_names.isdisjoint(kwargs.keys()):
+            raise TypeError(f"CircuitFunction {self.name} called with duplicate parameters.")
+        if len(args) + len(kwargs) != len(self._function_params):
+            raise TypeError(
+                f"CircuitFunction {self.name} takes {len(self.function_params)}"
+                f" parameters but {len(args) + len(kwargs)} were provided."
+            )
+        if not set(kwargs.keys()) <= expected_param_names:
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:136]: Due to the fact that this is tested after the length test above, this will only detect unrecognized keywords if the total argument count happens to match `len(self._function_params)` exactly. The convention in Python is to validate keyword args against expected names first.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+            raise TypeError(
+                f"CircuitFunction {self.name} called with unrecognized keyword arguments."
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:138]: To aid debugging, this should name the unrecognized arguments.

File: cirq-core/cirq/circuits/circuit_function.py [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+
+    def __call__(self, *args: cirq.TParamVal, **kwargs: cirq.TParamVal) -> cirq.FrozenCircuit:
+        """Call the circuit function with given parameters values."""
+
+        param_dict = dict(zip(self._function_params, args)) | kwargs
+        positional_param_names = {p.name for p in self._function_params[: len(args)]}
+        expected_param_names = {p.name for p in self._function_params}
+
+        if not positional_param_names.isdisjoint(kwargs.keys()):
+            raise TypeError(f"CircuitFunction {self.name} called with duplicate parameters.")
```
Comment (mhucka) [inline cirq-core/cirq/circuits/circuit_function.py:130]: To aid debugging, this should name the duplicate parameters.

**Review by mhucka — CHANGES_REQUESTED** [review]
Here are some preliminary review results.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 5 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Good catch, fixed.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 17 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py:91]: Added docstring.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 18 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py:96]: Added docstring.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 19 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py:101]: Added docstring.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 7 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Added detailed docstring.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 20 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Added more detail and specified the alphabetical default.

File: cirq-core/cirq/circuits/circuit_function_test.py (hunk 22 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function_test.py:164]: Good idea, added a check for FrozenCircuit.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 23 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: I moved this after the checks so an incorrect `param_dict` should never be formed now.

Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: `resolve_parameters` handles heterogeneous dicts, but I went ahead changed this to be a strict {string : TParamVal} dict anyways.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 24 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Great point, I had not considered that. I changed the order of the checks.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 25 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Good idea, unrecognized keyword arguments are named now.

File: cirq-core/cirq/circuits/circuit_function.py (hunk 26 above)
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function.py]: Good idea, duplicate parameters are named now.

File: cirq-core/cirq/circuits/circuit_function_test.py
Comment (arettig) [reply cirq-core/cirq/circuits/circuit_function_test.py:1]: Great suggestions, I added all of these except the last one.

For the last one, this testing function enforces that any parameters that your object has can be resolved. Since I don't want `CircuitFunction`s to be resolvable, I went ahead and removed `parameter_names` and `is_parameterized`, i.e. made it not a parametrizable object. The parameterization methods will be handled within `CallFunctionOperation`, the object that is actually part of the circuit.
